llama.cpp

```diff
@@ -7360,6 +7360,32 @@ void llama_sample_top_p(struct llama_context * ctx, llama_token_data_array * can
     }
 }
 
+void llama_sample_min_p(struct llama_context * ctx, llama_token_data_array * candidates, float p, size_t min_keep) {
+    if (p <= 0.0f || !candidates->size) {
```

Comment:
Same as my comment above. Imagine UI slider from 0 to 1, and when user sets it gradually from 1 to 0, user expects to get more deterministic predictions. But in this case (check for `p <= 0`), when min_p slider is 0, user will get unexpected behavior - 0 and 0.00001 behaves differently.

Reply:
Can you reproduce this discrepancy with the current code? I thought I understood what *p* was doing here, but you're telling me that larger values *already* let more tokens pass?

Reply:
I'm confused. This code block checks if it's equal to 0 and skips it if it is equal to zero (or is a negative value), which seems correct. The sampler should be more deterministic the closer it is to 1.0 in which 1.0 is the maximum.

Reply:
I think he was describing how he wants it to work (in line with existing samplers like TFS where you'd use a value like `0.95`, and to disable `1.0`) now how it currently does.

Reply:
Floats have this annoying property of not staying 0.0 when you want them to.

For example ( simplified to demonstrate the point ), this can fail:
```
float test = 0.0;
test += 0.123;
test += 0.123;
test -= 0.123;
test -= 0.123;
assert( test <= 0.0 );
```

Reply:
@staviq The way this code is currently supposed to work, min_p of e.g. 1e-9 should effectively let everything through as we could assume nothing has a probability that low - it's a minimum. I added this check so that 0.0 and below don't even need to run the sampling code.

Reply:
> I think he was describing how he wants it to work (in line with existing samplers like TFS where you'd use a value like `0.95`, and to disable `1.0`) not describing what it currently does.

Maybe it could be a integer scale from 0 - 100 instead? Where the number represents the actual %. So Min P (percentage?) of 5 would be 5% base requirement? Would this be more intuitive?

Reply:
Using a percent to represent a probability would be more confusing, I think. Seeing an integer above one might lead one to think it's like top-k and related to a number of tokens.

Reply:
> Using a percent to represent a probability would be more confusing, I think. Seeing an integer above one might lead one to think it's like top-k and related to a number of tokens.

Ah fair. Well I think how it exists right now is fine then in my opinion. 'Lower minimum means more is allowed' is intuitive to me